wp-content/plugins/tapestry/controller/class.user-controller.php

```diff
@@ -0,0 +1,141 @@
+<?php
+// TODO Change exceptions to using an ERROR class
+/**
+ * Add/update/retrieve User progress
+ * 
+ */
+class TapestryUserController {
+
+    /** 
+     * Update User's video progress for a tapestry post
+     * @param type @postId the post's ID
+     * @param type @nodeId the current node being viewed
+     * @param type @progressValue is how much the video was viewed, value should be between >= 0 and <= 1
+    */
+    public function updateProgress($postId, $nodeId, $progressValue) {
+        $userId = apply_filters('determine_current_user', false);
+        $this->checkUserAndPostId($userId, $postId);
+
+        if ($progressValue !== NULL) {
+            $progressValue = json_decode($progressValue);
+        }
+
+        // Value should be between 0 and 1
+        if ($progressValue < 0 || $progressValue > 1) {
+            throw new Exception('Invalid progress value');
+        }
+
+        $this->updateUserProgress($userId, $postId, $nodeId, $progressValue);
+    }
+
+    /** 
+     * Get User's video progress for a tapestry post
+     * @param type @postId the post's ID
+     * @param type @nodeIdArr is a list of ids currently in the tapestry
+    */
+    public function getProgress($postId, $nodeIdArr) {
+        $userId = apply_filters('determine_current_user', false);
+        $this->isPostValid($postId);
+        $this->checkUserAndPostId($userId, $postId);
+
+        if ($this->isJson($nodeIdArr)) {
+            $nodeIdArr = json_decode($nodeIdArr);
+        } else {
+            throw new Exception('Invalid json');
+        }
+
+        return $this->getUserProgress($userId, $postId, $nodeIdArr);
+    }
+
+    /** 
+     * Update User's h5p video setting for a tapestry post
+     * @param type @postId the post's ID
+     * @param type @h5pSettingsData stores volume, playbackRate, quality of h5p video
+    */
+    public function updateH5PSetting($postId, $h5pSettingsData) {
+        $userId = apply_filters('determine_current_user', false);
+        $this->checkUserAndPostId($userId, $postId);  
+
+        if ($this->isJson($h5pSettingsData)) {
+            $h5pSettingsData = json_decode($h5pSettingsData);
+        } else {
+            throw new Exception('Invalid json');
+        }
+
+        $this->updateUserH5PSetting($userId, $postId, $h5pSettingsData);
+    }
+
+    /** 
+     * Get User's h5p video setting for a tapestry post
+     * @param type @postId the post's Id
+    */
+    public function getH5PSetting($postId) {
+        $userId = apply_filters('determine_current_user', false);
+        $this->isPostValid($postId);
+        $this->checkUserAndPostId($userId, $postId);  
+        return $this->getUserH5PSetting($userId, $postId);
+    }
+
+    private function updateUserProgress($userId, $postId, $nodeId, $progressValue) {
```

Comment:
As a convention, please name all private functions starting with an underscore.

Reply:
Updated to include underscore before function name for private functions.